frontend/catalyst/compiled_functions.py

```diff
@@ -327,3 +333,119 @@ def __call__(self, *args, **kwargs):
         )
 
         return result
+
+
+@dataclass
+class CacheKey:
+    """A key by which to identify entries in the compiled function cache.
+
+    The cache only rembers one compiled function for each possible combination of:
+     - dynamic argument PyTree metadata
+     - static argument values
+    """
+
+    treedef: PyTreeDef
+    static_args: Tuple
+
+    def __hash__(self) -> int:
+        if not hasattr(self, "_hash"):
+            self._hash = hash((self.treedef, self.static_args))
+        return self._hash
+
+
+@dataclass
+class CacheEntry:
+    """An entry in the compiled function cache.
+
+    For each compiled function, the cache stores the dynamic argument signature, the output PyTree
+    definition, as well as the workspace in which compilation takes place.
+    """
+
+    compiled_fn: CompiledFunction
+    signature: Tuple
+    out_treedef: PyTreeDef
+    workspace: Directory
+
+
+class CompilationCache:
+    """Class to manage CompiledFunction instances and retrieving previously compiled versions of
+    a function.
+
+    A full match requires the following properties to match:
+     - dynamic argument signature (the shape and dtype of flattened arrays)
+     - dynamic argument PyTree definitions
+     - static argument values
+
+    In order to allow some flexibility in the type of arguments provided by the user, a match is
+    also produced if the dtype of dynamic arguments can be promoted to the dtype of an existing
+    signature via JAX type promotion rules. Additional leniency is provided in the shape of
+    dynamic arguments in accordance with the abstracted axis specification.
+
+    To avoid excess promotion checks, only one function version is stored at a time for a given
+    combination of PyTreeDefs and static arguments.
+    """
+
+    def __init__(self, static_argnums, abstracted_axes):
+        self.static_argnums = static_argnums
+        self.abstracted_axes = abstracted_axes
+        self.cache = {}
+
+    def get_function_status_and_key(self, args):
```

Comment:
Good point, the function uses the TypeCompatibility enum as a stand in for whether there was a full or partial cache hit, or a cache miss. It's not super ideal, but I think I'd rather the function describe what it is intended to _do_ / how it is intended to be used, rather than focus on it's implementation. I have updated the docstring though to make it clear what exactly it returns.